library/Backends/llvm/src/Juvix/Backends/LLVM/Codegen/Closure.hs

```diff
@@ -0,0 +1,149 @@
+module Juvix.Backends.LLVM.Codegen.Closure
+  ( register,
+    reference,
+    environmentPtr,
+    environmentPtrDeref,
+    functionPtr,
+    mallocEnvironment,
+    malloc,
+    pointer,
+
+    -- * Pointer Storage
+    storeFunctionPtr,
+    storeEnvironmentPtr,
+
+    -- * Storage Lookup
+    getFunctionPtr,
+    loadFunctionPtr,
+    getEnvironmentPtr,
+    loadEnvironmentPtr,
+  )
+where
+
+import qualified Juvix.Backends.LLVM.Codegen.Block as Block
+import qualified Juvix.Backends.LLVM.Codegen.Types as Types
+import Juvix.Library hiding (Type, local)
+import qualified LLVM.AST as AST
+import qualified LLVM.AST.Type as Type
+
+-- | @register@ registers the type with the name closure
+register :: HasState "moduleDefinitions" [AST.Definition] m => m ()
+register =
+  Block.addType name type'
+
+name :: IsString p => p
+name = "closure"
+
+pointer :: Type.Type
+pointer = Types.pointerOf reference
+
+-- | @reference@ the name to reference the closure type, this is what
+-- one ought to use when referencing the type.
+reference :: Type.Type
+reference = Type.NamedTypeReference name
+
+-- | @type'@ the base type of a closure
+type' :: Type.Type
+type' =
+  Type.StructureType
+    { isPacked = True,
+      elementTypes =
+        [ -- function pointer
+          -- IMPORTANT Keep at the head so we can deference it with no
+          -- effort
+          functionPtr,
+          -- the closure array
+          environmentPtr
+        ]
+    }
+
+sizeOf :: Num a => a
+sizeOf = Types.addressSpace * 2
+
+malloc :: Types.Call m => m AST.Operand
+malloc = Block.malloc sizeOf pointer
+
+mallocEnvironment :: Types.Call m => Integer -> m AST.Operand
+mallocEnvironment length =
+  -- is this the correct offset to get it to compile
```

Comment:
Is this question now answered in the affirmative?  :)

Reply:
seems so!